### PR TITLE
port base64 encoding from 3.4

### PR DESCRIPTION
**Pull request by rogday**
issue #14977

Test data is at https://github.com/opencv/opencv_extra/pull/877

```
opencv_extra=base64_encoding
force_builders=Custom,ARMv7
build_image:Custom=centos:7
buildworker:Custom=linux-1
```

### CODE REVIEW

File: modules/core/src/persistence.cpp [hunk 1]
```diff
@@ -339,1449 +341,1463 @@ static inline void writeReal(uchar* p, double fval)
...
+    size_t write_stack_size = write_stack.size();
+    if (write_stack_size > 1)
+        write_stack[write_stack_size - 2].flags &= ~FileNode::EMPTY;
 
-        if( fmt == FileStorage::FORMAT_JSON && type_name && type_name[0] && FileNode::isMap(struct_flags))
-        {
-            emitter->write("type_id", type_name, false);
-        }
-    }
+    if (fmt != FileStorage::FORMAT_JSON && !FileNode::isFlow(s.flags))
```
Comment (rogday) [inline modules/core/src/persistence.cpp:949]: This somehow fixes JSON indentation mismatch with 3.4, but looks fragile.

File: modules/core/src/persistence_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,228 @@
+#ifndef OPENCV_CORE_PERSISTENCE_IMPL_HPP
+#define OPENCV_CORE_PERSISTENCE_IMPL_HPP
+
+#include "precomp.hpp"
```
Comment (alalek) [inline modules/core/src/persistence_impl.hpp:4]: > precomp.hpp

must be used in .cpp files only (as unconditional first include)

File: modules/python/test/test_filestorage_io.py [hunk 3]
```diff
@@ -109,5 +111,92 @@ def test_yml(self):
...
+        os.close(fd)
+        np.random.seed(42)
+        self.write_base64_json(fname)
+        os.remove(fname)
+
+    @staticmethod
+    def get_normal_2d_mat():
+        rows = 10
+        cols = 20
+        ch = 3
```
Comment (alalek) [inline modules/python/test/test_filestorage_io.py:125]: > ch

`cn` is usually used in OpenCV code.

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

File: modules/core/src/base64_encoding.cpp [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+        }
+    }
+
+    ~Base64ContextEmitter()
+    {
+        /* cleaning */
+        if (src_cur != src_beg)
+            flush();    /* encode the rest binary data to base64 buffer */
+
+        if ( file_storage->fmt == cv::FileStorage::Mode::FORMAT_JSON )
```
[the remark this answers is not in the document]
Comment (rogday) [reply modules/core/src/base64_encoding.cpp]: I removed JSON-specific actions from constructor and destructor( writing `"$base64` and `"` to buffer) and added need_indent field instead.

File: modules/core/src/base64_encoding.cpp [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+
+base64::Base64Writer::~Base64Writer()
+{
+    delete emitter;
+}
+
+void base64::Base64Writer::check_dt(const char* dt)
+{
+    if ( dt == 0 )
+        CV_Error( CV_StsBadArg, "Invalid \'dt\'." );
```
[the remark this answers is not in the document]
Comment (rogday) [reply modules/core/src/base64_encoding.cpp]: Fixed

File: modules/python/test/test_filestorage_io.py (hunk 3 above)
Comment (rogday) [reply modules/python/test/test_filestorage_io.py]: Fixed

File: modules/core/src/persistence_impl.hpp (hunk 2 above)
Comment (rogday) [reply modules/core/src/persistence_impl.hpp]: Fixed

File: modules/core/src/base64_encoding.cpp [hunk 6]
```diff
@@ -0,0 +1,274 @@
+#include "persistence_impl.hpp"
+#include "base64_encoding.hpp"
+
+class base64::Base64ContextEmitter
+{
+public:
+    explicit Base64ContextEmitter(cv::FileStorage::Impl * fs)
```
[the remark this answers is not in the document]
Comment (rogday) [reply modules/core/src/base64_encoding.cpp]: As you suggested, I replaced pointer to reference.

File: modules/core/src/base64_encoding.cpp [hunk 7]
```diff
@@ -0,0 +1,274 @@
+#include "persistence_impl.hpp"
+#include "base64_encoding.hpp"
+
+class base64::Base64ContextEmitter
```
[the remark this answers is not in the document]
Comment (rogday) [reply modules/core/src/base64_encoding.cpp]: Thank you for review. Fixed

**Comment by alalek** [comment]
Is this PR still in WIP status? (as per PR title)

**Comment by rogday** [comment]
> Is this PR still in WIP status? (as per PR title)

I squashed commits, changed the title, it's ready for merge.

**Review by asmorkalov — APPROVED** [review]

**Pull request merged**